tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1306 @@
+
+  intrusive_ptr_increment(state.get());  // adopted by ListTask::Start.
+  read_rate_limiter().Admit(state.get(), &ListTask::Start);
+}
+
+// Receiver used by `DeleteRange` for processing the results from `List`.
+struct DeleteRangeListReceiver {
+  IntrusivePtr<S3KeyValueStore> owner_;
+  Promise<void> promise_;
+  FutureCallbackRegistration cancel_registration_;
+
+  void set_starting(AnyCancelReceiver cancel) {
+    cancel_registration_ = promise_.ExecuteWhenNotNeeded(std::move(cancel));
+  }
+
+  void set_value(std::string key) {
+    assert(!key.empty());
+    if (!key.empty()) {
+      LinkError(promise_, owner_->Delete(std::move(key)));
+    }
+  }
+
+  void set_error(absl::Status error) {
+    SetDeferredResult(promise_, std::move(error));
+    promise_ = Promise<void>();
+  }
+
+  void set_done() { promise_ = Promise<void>(); }
+
+  void set_stopping() { cancel_registration_.Unregister(); }
+};
+
+Future<const void> S3KeyValueStore::DeleteRange(KeyRange range) {
+  s3_delete_range.Increment();
+  if (range.empty()) return absl::OkStatus();
+
+  // TODO(jbms): It could make sense to rate limit the list operation, so that
+  // we don't get way ahead of the delete operations.  Currently our
+  // sender/receiver abstraction does not support back pressure, though.
+  auto op = PromiseFuturePair<void>::Make(tensorstore::MakeResult());
+  ListOptions list_options;
+  list_options.range = std::move(range);
+  ListImpl(list_options, DeleteRangeListReceiver{
+                             internal::IntrusivePtr<S3KeyValueStore>(this),
+                             std::move(op.promise)});
+  return std::move(op.future);
+}
+
+Result<kvstore::Spec> ParseS3Url(std::string_view url) {
+  auto parsed = internal::ParseGenericUri(url);
+  assert(parsed.scheme == kUriScheme);
+  if (!parsed.query.empty()) {
+    return absl::InvalidArgumentError("Query string not supported");
+  }
+  if (!parsed.fragment.empty()) {
+    return absl::InvalidArgumentError("Fragment identifier not supported");
+  }
+  size_t end_of_bucket = parsed.authority_and_path.find('/');
+  std::string_view bucket = parsed.authority_and_path.substr(0, end_of_bucket);
+  if (!IsValidBucketName(bucket)) {
+    return absl::InvalidArgumentError(
+        tensorstore::StrCat("Invalid S3 bucket name: ", QuoteString(bucket)));
+  }
+  std::string_view path =
+      (end_of_bucket == std::string_view::npos)
+          ? std::string_view{}
+          : parsed.authority_and_path.substr(end_of_bucket + 1);
+  auto driver_spec = internal::MakeIntrusivePtr<S3KeyValueStoreSpec>();
+  driver_spec->data_.bucket = bucket;
+  driver_spec->data_.request_concurrency =
+      Context::Resource<S3ConcurrencyResource>::DefaultSpec();
+  driver_spec->data_.retries =
+      Context::Resource<S3RequestRetries>::DefaultSpec();
+  driver_spec->data_.data_copy_concurrency =
+      Context::Resource<DataCopyConcurrencyResource>::DefaultSpec();
+
+  driver_spec->data_.requester_pays = true;
```

Comment:
Should be `false`, I think --- this should have explicit acknowledgement.

Reply:
done